---
title: "Recoding SNPs"
author: "Felicia New"
date: "4/17/2019"
output: pdf_document
---


```{r, echo=F}
library(data.table)

snp <- as.data.frame(fread("~/Docs/brito_lab/twins_analysis/snp/BGI_twins_maf10_nomissing_formatAnalysis.csv", header=T, sep=','))
  rownames(snp) <- snp[,1]
  snp[,1] <- NULL
 
recode <- function(c) {
  d <- matrix(nrow = length(c), ncol = 2)
  rownames(d) <- rownames(snp)
  for (i in 1:length(c)) {
    if (c[i] == 0) {
      d[i,] <- c(0,0)
    } else if (c[i] == 1) {
      d[i,] <- c(1,0)
    } else {
      d[i,] <- c(0,1)
    }
  }
  return(d)
}

mats <- lapply(snp, recode)

mat2 <- matrix(nrow = nrow(snp), ncol = 2*ncol(snp))
rownames(mat2) <- rownames(snp)
for (j in 1:(ncol(snp))) {
  mat2[,(2*j-1):(2*j)] <- mats[[j]]
}

#I need to modify this table....I need column and row names
write.table(mat2, "~/Docs/brito_lab/twins_analysis/snp/BGI_twins_maf10_nomissing_dummy.csv", sep=",", row.names = T, quote=F, col.names = F)
```

##Remove the ten people without SNPs from the gene abundance file.
```{r, echo=F}
gene <- as.data.frame(fread("~/Docs/brito_lab/twins_analysis/twins_function_abundance_filt25.csv", header=T, sep=',')) 

gene[is.na(gene)] <- 0 #Set NAs to 0.
rownames(gene) <- gene[,1]
gene[,1] <- NULL

names <- rownames(snp)
gene.sub <- subset(gene, rownames(gene) %in% names)

#Save this file.
write.csv(gene.sub,"~/Docs/brito_lab/twins_analysis/twins240_function_abundance_filt25.csv", quote=F,row.names=T)
```


```{r,echo=F}
#Plotting the parallel results
cores <- c(2,4,6,8,10,12,14,16)
seconds <- c(350.456,237.554,153.379,159.254,112.541,110.727,121.025,74.666)
pdf("~/Docs/brito_lab/twins_analysis/scca/parallel_computing_time.pdf")
plot(cores, seconds, pch=20)
dev.off()
```


```{r}
#meta <- read.csv("~/Docs/brito_lab/twins_analysis/twinsUK_zyg_age_ship_family.csv", header=T, row.names=1)
#meta1 <- meta[,c(-5)]
#meta240 <- subset(meta, rownames(meta) %in% names)

#meta.names <- rownames(meta1)
#meta240 <- subset(meta1, rownames(meta1) %in% bgi)
'%ni%' <- Negate('%in%')
#lost<-subset(meta1, meta.names %ni% names)

#This metadata is missing 4 individuals. I don't know when it lost them, so I need to remake this file.
#meta <- read.csv("~/Docs/brito_lab/twins_metadata/Brito_BGI_twins_metadata.csv")
#  meta$BGI_ID <- sub("^", "BGI", meta$BGI_ID)
#  meta.short <- meta[,c(1,5,12,13,27,17)]
#meta.short$MZTwin <- 0
#meta.short$MZTwin[meta.short$IndividualZygosity=='MZ'] <-1

#meta.short2 <- meta.short[,c(-3)]
#rownames(meta.short2) <- meta.short2[,1]
#meta.short2[,1] <- NULL
#meta.final <- meta.short2

#SAve this file. 
write.csv(meta.final, "~/Docs/brito_lab/twins_analysis/twinsUK_zyg_age_ship_bmi_family.csv", quote=F)
#Save a version with 240/250
meta.240 <- subset(meta.final, rownames(meta.final) %in% names)
write.csv(meta.240, "~/Docs/brito_lab/twins_analysis/twinsUK240_zyg_age_ship_bmi_family.csv", quote=F)

#This is the final metadata for the 240 set
meta.240 <- read.csv("~/Docs/brito_lab/twins_analysis/twinsUK240_zyg_age_ship_bmi_family.csv", header=T, row.names=1)
```


```{r}
#Check why matrix.model is throwing errors on the server.
X1.a <- as.data.frame(fread("~/Docs/brito_lab/twins_analysis/snp/BGI_twins_maf10_nomissing_dummy.csv", header=F, sep=','))
rownames(X1.a) <- X1.a[,1]
X1.a[,1] <- NULL

X1.as <- as.matrix(X1.a[ order(row.names(X1.a)),])

mod <- svd(X1.as)
pc <- mod$u[,1] * mod$d[1]

A <- as.matrix(model.matrix(~pc + Age.at.metagenomics.sample + + MZTwin + SampleShipmentNum + IndividualBMI + IndividualFamilyID, data = meta.240))
A2 <- crossprod(A,A)
AX1 <- crossprod(A, X1.as)
a2ax1 <- ginv(A2, AX1)
X1 <- X1.as - A %*% a2ax1
X1 <- (X1.as) - A %*% solve(A2, AX1)


write.csv(A, "~/Docs/brito_lab/twins_analysis/metadata_model_matrix_output.csv", quote=F, row.names=F)
```


```{r}
#Do PCA on the SNP table and look at singular values for Ben
#Demean the SNPs
snp.demean <- apply(X1.as, 2, FUN= function(col) col - mean(col))
mod <- svd(snp.demean)

mod$d

pc <- mod$u[,1] * mod$d[1]
snp.pca <- prcomp(X1.as)
snp.pca$sdev
pdf("~/Docs/brito_lab/twins_analysis/snp/snp_pca_plot_var.pdf")
plot(snp.pca$sdev^2, ylab=c("Variance"),main=c("SNP PCA"))
dev.off()

plot(snp.pca$sdev^2/sum(snp.pca$sdev^2)*100)
forben <- dput(snp.pca$sdev)

#PCA on additive snp table
add.pca <- prcomp(snp)
plot(add.pca$sdev^2)

snp.matrix <- as.matrix(snp)
adddom.pca <- prcomp(cbind(snp.matrix-1, 1-2*abs(snp.matrix-1)))

plot(adddom.pca$x[,1], adddom.pca$x[,2], xlim=c(-50,50), ylim=c(-50,50))
plot(jitter(adddom.pca$x[,1])+rnorm(240, 1/2), adddom.pca$x[,2] + rnorm(240, 1/2), xlim=c(-50,50), ylim=c(-50,50))
plot(adddom.pca$x[,1], adddom.pca$x[,2], xlim=c(-50,50), ylim=c(-50,50))
plot(adddom.pca$x[,1], adddom.pca$x[,2])

#What happens if we subsample?
subsample <- sample(1:ncol(snp.matrix), 10000)
snp.mat.red <- snp.matrix[,subsample]
adddom.red.pca <- prcomp(cbind(snp.mat.red-1, 1-2*abs(snp.mat.red-1)))
plot(adddom.red.pca$sdev^2/sum(adddom.red.pca$sdev^2))

#Remove every other MZ twin (keep only one set of the MZ data for each pair)
mz.twins <- which(meta.240[,5] == 1)
mz.twins <- mz.twins[mz.twins!=142]
mz.twin <- mz.twins[2*(1:(length(mz.twins)/2))-1]
snp.red <- snp[-mz.twin,]

#PCA on the reduced dataset.
adddom.red.pca <- prcomp(cbind(snp.red-1, 1-2*abs(snp.red-1)))
plot(adddom.red.pca$x[,1], adddom.red.pca$x[,2])
plot(adddom.red.pca$sdev^2)
red.pca.x <- adddom.red.pca$x
write.csv(red.pca.x, "~/Docs/brito_lab/twins_analysis/snp/reduced_snp_pca_x.csv", quote=F)

#It looks like the repeated rows of the MZ twins are causing one issue.
snp.pca.x <- snp.pca$x
write.csv(snp.pca.x, "~/Docs/brito_lab/twins_analysis/snp/full_snp_pca_x.csv", quote=F)
```


```{r, echo=F}
#Remove one from each family, remove first DZ twin, dont forget the lonely twins
#NA5172 NA5369 NA5937 NA5530 NA9465 NA9457 NA9419 NA9229
#First get dataframe of the lonely twins metadata. 
#Then remove the lonely twins from the metadata. Remove every other row. then add the lonely twins back. now subset the snp table with this.
names <- c("NA5172", "NA5369", "NA5937", "NA5530", "NA9465", "NA9457", "NA9419", "NA9229")

lonely <- (meta.240)[((meta.240$IndividualFamilyID) %in% names),]
pair <- meta.240[!((meta.240$IndividualFamilyID) %in% names),]

toDelete <- seq(1, nrow(pair), 2)
singles <- pair[ toDelete ,]

all.single <- rbind(singles, lonely) #124
tokeep <- rownames(all.single)
snp.sing <- subset(X1.as, rownames(X1.as) %in% tokeep)

#Do PCA
pca.single <- prcomp(snp.sing)
meta.240$color="black"
meta.240$color[meta.240$MZTwin==1]="red"
plot(pca.single$x[,1], pca.single$x[,2], col=meta.240$color, main=c("PCA with one twin from each pair\n red=MZ, black=DZ"), ylab=c("PC2"), xlab=c("PC1"))


pdf("~/Docs/brito_lab/twins_analysis/scca/pca/pca_variance_plot_one_from_each_pair.pdf")
plot(pca.single$sdev^2/sum(pca.single$sdev^2), main=c("PCA with only one from each pair\n124 individuals"), ylab=c("Proportion of variance"))
dev.off()
```


##I pulled out the full twin set that Emily sent. I have the SNP table but I need to make sure I have zygosity and familyID? Maybe I don't need those for the PCA...
```{r, echo=F}
full.snp <- as.data.frame(fread("~/Docs/brito_lab/twins_analysis/snp/twinsuk2476_nomiss_maf5_format4analysis.csv", header=T, sep=','))
  row.names(full.snp) <- full.snp[,1]
  full.snp[,1] <- NULL

full.mat <- as.matrix(full.snp)
full.snp.pca <- prcomp(cbind(full.mat-1, 1-2*abs(full.mat-1)))

```

##Plotting the abundances just to look...
```{r}
#Plot the gene abundances
plot(density(as.numeric(gene[1:nrow(gene),])))
gden <- apply(gene, 1, density)

densities <- apply(gene, 2, density)

plot(densities[[1]], ylim=c(0,20), xlim=c(0,4), main="Gene coverage is not normal")
for(i in 2:length(densities)){
  lines(densities[[i]], col=i)
}

samp <- apply(gene, 1, density)

plot(samp[[1]], ylim=c(0,1), xlim=c(0,200), main="Samples are not normal")
for(i in 2:length(samp)){
  lines(samp[[i]], col=i)
}

```